// @neo4j/graphql/dist/classes/Executor.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Executor = void 0;
const neo4j_driver_1 = require("neo4j-driver");
const debug_1 = __importDefault(require("debug"));
const environment_1 = __importDefault(require("../environment"));
const Error_1 = require("./Error");
const constants_1 = require("../constants");
const create_auth_param_1 = __importDefault(require("../translate/create-auth-param"));
const debug = (0, debug_1.default)(constants_1.DEBUG_EXECUTE);
function isDriverLike(executionContext) {
    return typeof executionContext.session === "function";
}
function isSessionLike(executionContext) {
    return typeof executionContext.beginTransaction === "function";
}
class Executor {
    constructor({ executionContext, auth, queryOptions, database, bookmarks }) {
        this.executionContext = executionContext;
        this.lastBookmark = null;
        this.queryOptions = queryOptions;
        if (auth) {
            this.auth = auth;
        }
        else {
            this.auth = (0, create_auth_param_1.default)({ context: {} });
        }
        this.database = database;
        this.bookmarks = bookmarks;
    }
    async execute(query, parameters, defaultAccessMode) {
        try {
            if (isDriverLike(this.executionContext)) {
                const session = this.executionContext.session(this.getSessionParam(defaultAccessMode));
                const result = await this.sessionRun(query, parameters, defaultAccessMode, session);
                await session.close();
                return result;
            }
            if (isSessionLike(this.executionContext)) {
                return await this.sessionRun(query, parameters, defaultAccessMode, this.executionContext);
            }
            return await this.transactionRun(query, parameters, this.executionContext);
        }
        catch (error) {
            throw this.formatError(error);
        }
    }
    formatError(error) {
        if (error instanceof neo4j_driver_1.Neo4jError) {
            if (error.message.includes(`Caused by: java.lang.RuntimeException: ${constants_1.AUTH_FORBIDDEN_ERROR}`)) {
                return new Error_1.Neo4jGraphQLForbiddenError("Forbidden");
            }
            if (error.message.includes(`Caused by: java.lang.RuntimeException: ${constants_1.AUTH_UNAUTHENTICATED_ERROR}`)) {
                return new Error_1.Neo4jGraphQLAuthenticationError("Unauthenticated");
            }
            if (error.message.includes(`Caused by: java.lang.RuntimeException: ${constants_1.RELATIONSHIP_REQUIREMENT_PREFIX}`)) {
                const [, message] = error.message.split(constants_1.RELATIONSHIP_REQUIREMENT_PREFIX);
                return new Error_1.Neo4jGraphQLRelationshipValidationError(message);
            }
            if (error.code === "Neo.ClientError.Schema.ConstraintValidationFailed") {
                return new Error_1.Neo4jGraphQLConstraintValidationError("Constraint validation failed");
            }
        }
        debug("%s", error);
        return error;
    }
    generateQuery(query) {
        if (this.queryOptions && Object.keys(this.queryOptions).length) {
            const queryOptions = `CYPHER ${Object.entries(this.queryOptions)
                .map(([key, value]) => `${key}=${value}`)
                .join(" ")}`;
            return `${queryOptions}\n${query}`;
        }
        return query;
    }
    generateParameters(query, parameters) {
        if (query.includes("$auth.") || query.includes("auth: $auth") || query.includes("auth:$auth")) {
            return { ...parameters, auth: this.auth };
        }
        return parameters;
    }
    getSessionParam(defaultAccessMode) {
        // Always specify a default database to avoid requests for routing table
        const sessionParam = { defaultAccessMode, database: "neo4j" };
        if (this.database) {
            sessionParam.database = this.database;
        }
        if (this.bookmarks) {
            sessionParam.bookmarks = this.bookmarks;
        }
        return sessionParam;
    }
    getTransactionConfig() {
        const app = `${environment_1.default.NPM_PACKAGE_NAME}@${environment_1.default.NPM_PACKAGE_VERSION}`;
        return {
            metadata: {
                app,
                type: "user-transpiled",
            },
        };
    }
    async sessionRun(query, parameters, defaultAccessMode, session) {
        const transactionType = `${defaultAccessMode.toLowerCase()}Transaction`;
        const result = await session[transactionType]((transaction) => {
            return this.transactionRun(query, parameters, transaction);
        }, this.getTransactionConfig());
        const lastBookmark = session.lastBookmark();
        if (Array.isArray(lastBookmark) && lastBookmark[0]) {
            this.lastBookmark = lastBookmark[0];
        }
        return result;
    }
    transactionRun(query, parameters, transaction) {
        const queryToRun = this.generateQuery(query);
        const parametersToRun = this.generateParameters(query, parameters);
        debug("%s", `About to execute Cypher:\nCypher:\n${queryToRun}\nParams:\n${JSON.stringify(parametersToRun, null, 2)}`);
        return transaction.run(queryToRun, parametersToRun);
    }
}
exports.Executor = Executor;
