// @neo4j/cypher-builder/dist/expressions/functions/CypherFunction.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.randomUUID = exports.sum = exports.avg = exports.max = exports.min = exports.count = exports.cypherTime = exports.cypherLocalTime = exports.cypherLocalDatetime = exports.cypherDate = exports.cypherDatetime = exports.labels = exports.pointDistance = exports.distance = exports.point = exports.coalesce = exports.CypherFunction = void 0;
const CypherASTNode_1 = require("../../CypherASTNode");
class CypherFunction extends CypherASTNode_1.CypherASTNode {
    constructor(name, params = []) {
        super();
        this.name = name;
        this.params = params;
        for (const param of params) {
            if (param instanceof CypherASTNode_1.CypherASTNode) {
                this.addChildren(param);
            }
        }
    }
    getCypher(env) {
        const argsStr = this.params.map((expr) => expr.getCypher(env)).join(", ");
        return `${this.name}(${argsStr})`;
    }
}
exports.CypherFunction = CypherFunction;
function coalesce(expr, ...optionalExpr) {
    return new CypherFunction("coalesce", [expr, ...optionalExpr]);
}
exports.coalesce = coalesce;
function point(variable) {
    return new CypherFunction("point", [variable]);
}
exports.point = point;
function distance(lexpr, rexpr) {
    return new CypherFunction("distance", [lexpr, rexpr]);
}
exports.distance = distance;
function pointDistance(lexpr, rexpr) {
    return new CypherFunction("point.distance", [lexpr, rexpr]);
}
exports.pointDistance = pointDistance;
function labels(nodeRef) {
    return new CypherFunction("labels", [nodeRef]);
}
exports.labels = labels;
function cypherDatetime() {
    return new CypherFunction("datetime");
}
exports.cypherDatetime = cypherDatetime;
// TODO: Add optional input to date functions - https://neo4j.com/docs/cypher-manual/current/functions/#header-query-functions-temporal-instant-types
function cypherDate() {
    return new CypherFunction("date");
}
exports.cypherDate = cypherDate;
function cypherLocalDatetime() {
    return new CypherFunction("localdatetime");
}
exports.cypherLocalDatetime = cypherLocalDatetime;
function cypherLocalTime() {
    return new CypherFunction("localtime");
}
exports.cypherLocalTime = cypherLocalTime;
function cypherTime() {
    return new CypherFunction("time");
}
exports.cypherTime = cypherTime;
function count(expr) {
    return new CypherFunction("count", [expr]);
}
exports.count = count;
function min(expr) {
    return new CypherFunction("min", [expr]);
}
exports.min = min;
function max(expr) {
    return new CypherFunction("max", [expr]);
}
exports.max = max;
function avg(expr) {
    return new CypherFunction("avg", [expr]);
}
exports.avg = avg;
function sum(expr) {
    return new CypherFunction("sum", [expr]);
}
exports.sum = sum;
function randomUUID() {
    return new CypherFunction("randomUUID");
}
exports.randomUUID = randomUUID;
