// @neo4j/graphql/dist/schema/resolvers/query/root-connection.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.rootConnectionResolver = void 0;
const graphql_compose_1 = require("graphql-compose");
const utils_1 = require("../../../utils");
const translate_1 = require("../../../translate");
const get_neo4j_resolve_tree_1 = __importDefault(require("../../../utils/get-neo4j-resolve-tree"));
const utils_2 = require("../../../utils/utils");
const pagination_1 = require("../../pagination");
const fulltext_1 = require("../../../schema/augment/fulltext");
function rootConnectionResolver({ node, composer }) {
    async function resolve(_root, args, _context, info) {
        const context = _context;
        const resolveTree = (0, get_neo4j_resolve_tree_1.default)(info);
        const edgeTree = resolveTree.fieldsByTypeName[`${(0, graphql_compose_1.upperFirst)(node.plural)}Connection`].edges;
        const nodeTree = edgeTree?.fieldsByTypeName[`${node.name}Edge`].node;
        const resolveTreeForContext = nodeTree || resolveTree;
        context.resolveTree = { ...resolveTreeForContext, args: resolveTree.args };
        const { cypher, params } = (0, translate_1.translateRead)({ context, node, isRootConnectionField: true });
        const executeResult = await (0, utils_1.execute)({
            cypher,
            params,
            defaultAccessMode: "READ",
            context,
        });
        let totalCount = 0;
        let edges = [];
        let pageInfo = {
            hasNextPage: false,
            hasPreviousPage: false,
            startCursor: null,
            endCursor: null,
        };
        if (executeResult.records[0]) {
            const record = executeResult.records[0].this;
            totalCount = (0, utils_2.isNeoInt)(record.totalCount) ? record.totalCount.toNumber() : record.totalCount;
            const connection = (0, pagination_1.createConnectionWithEdgeProperties)({
                selectionSet: resolveTree,
                source: { edges: record.edges },
                args: { first: args.first, after: args.after },
                totalCount,
            });
            edges = connection.edges;
            pageInfo = connection.pageInfo;
        }
        return {
            totalCount,
            edges,
            pageInfo,
        };
    }
    const rootEdge = composer.createObjectTC({
        name: `${node.name}Edge`,
        fields: {
            cursor: "String!",
            node: `${node.name}!`,
        },
    });
    const rootConnection = composer.createObjectTC({
        name: `${(0, graphql_compose_1.upperFirst)(node.plural)}Connection`,
        fields: {
            totalCount: "Int!",
            pageInfo: "PageInfo!",
            edges: rootEdge.NonNull.List.NonNull,
        },
    });
    // since sort is not created when there is nothing to sort, we check for its existence
    let sortArg;
    if (composer.has(`${node.name}Sort`)) {
        sortArg = composer.getITC(`${node.name}Sort`);
    }
    return {
        type: rootConnection.NonNull,
        resolve,
        args: {
            first: "Int",
            after: "String",
            where: `${node.name}Where`,
            ...(sortArg ? { sort: sortArg.List } : {}),
            ...(node.fulltextDirective
                ? {
                    fulltext: {
                        type: `${node.name}Fulltext`,
                        directives: [
                            {
                                name: "deprecated",
                                args: {
                                    reason: fulltext_1.fulltextArgDeprecationMessage,
                                },
                            },
                        ],
                    },
                }
                : {}),
        },
    };
}
exports.rootConnectionResolver = rootConnectionResolver;
