// @neo4j/graphql/dist/graphql/directives/id.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.idDirective = void 0;
const graphql_1 = require("graphql");
exports.idDirective = new graphql_1.GraphQLDirective({
    name: "id",
    description: "Indicates that the field is an identifier for the object type. By default; autogenerated, and has a unique node property constraint in the database.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        autogenerate: {
            defaultValue: false,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLBoolean),
        },
        unique: {
            defaultValue: true,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLBoolean),
        },
        global: {
            description: "Opt-in to implementing the Node interface with a globally unique id",
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLBoolean),
            defaultValue: false,
        },
    },
});
