// @neo4j/graphql/dist/graphql/directives/timestamp.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.timestampDirective = void 0;
const graphql_1 = require("graphql");
const TimestampOperation_1 = require("./arguments/enums/TimestampOperation");
exports.timestampDirective = new graphql_1.GraphQLDirective({
    name: "timestamp",
    description: "Instructs @neo4j/graphql to generate timestamps on particular events, which will be available as the value of the specified field.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        operations: {
            description: "Which events to generate timestamps on. Defaults to both create and update.",
            defaultValue: TimestampOperation_1.TimestampOperationEnum.getValues().map((v) => v.value),
            type: new graphql_1.GraphQLNonNull(new graphql_1.GraphQLList(new graphql_1.GraphQLNonNull(TimestampOperation_1.TimestampOperationEnum))),
        },
    },
});
