// @neo4j/graphql/dist/graphql/scalars/index.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphQLTime = exports.GraphQLLocalTime = exports.GraphQLLocalDateTime = exports.GraphQLDuration = exports.GraphQLDateTime = exports.GraphQLDate = exports.GraphQLBigInt = void 0;
var BigInt_1 = require("./BigInt");
Object.defineProperty(exports, "GraphQLBigInt", { enumerable: true, get: function () { return BigInt_1.GraphQLBigInt; } });
var Date_1 = require("./Date");
Object.defineProperty(exports, "GraphQLDate", { enumerable: true, get: function () { return Date_1.GraphQLDate; } });
var DateTime_1 = require("./DateTime");
Object.defineProperty(exports, "GraphQLDateTime", { enumerable: true, get: function () { return DateTime_1.GraphQLDateTime; } });
var Duration_1 = require("./Duration");
Object.defineProperty(exports, "GraphQLDuration", { enumerable: true, get: function () { return Duration_1.GraphQLDuration; } });
var LocalDateTime_1 = require("./LocalDateTime");
Object.defineProperty(exports, "GraphQLLocalDateTime", { enumerable: true, get: function () { return LocalDateTime_1.GraphQLLocalDateTime; } });
var LocalTime_1 = require("./LocalTime");
Object.defineProperty(exports, "GraphQLLocalTime", { enumerable: true, get: function () { return LocalTime_1.GraphQLLocalTime; } });
var Time_1 = require("./Time");
Object.defineProperty(exports, "GraphQLTime", { enumerable: true, get: function () { return Time_1.GraphQLTime; } });
